### PR TITLE
OHE tool wrapper

### PR DESCRIPTION
**Pull request by kxk302**
This is for https://github.com/bgruening/galaxytools/issues/1093

I wrapped the OHE tool by looking at how existing Keras tools are wrapped. Added tests too. However, I am not clear how to run the tests/tool. Could you please review the change and also point me to any documentation on how to test a wrapped tool in ML workbench?

Thanks!

### CODE REVIEW

File: tools/sklearn/test-data/ohe_out_5.tsv [hunk 1]
```diff
@@ -0,0 +1,8 @@
+1	0	0	0	0
```
Comment (anuprulez) [inline tools/sklearn/test-data/ohe_out_5.tsv:1]: I think the last column of zeros is because the `num_classes` parameter is set to `5`?

**Comment by qiagu** [comment]
Have you tried OneHotEncoder in scikit-learn? 
https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html#sklearn.preprocessing.OneHotEncoder
If that works, adding an option at the position below seems to an simple solution.
https://github.com/bgruening/galaxytools/blob/master/tools/sklearn/main_macros.xml#L788

File: tools/sklearn/test-data/ohe_out_5.tsv (hunk 1 above)
Comment (kxk302) [reply tools/sklearn/test-data/ohe_out_5.tsv:1]: Yes, this is a test where we have 4 classes in the input file, but num_classes is intentionally set to 5.

**Comment by kxk302** [comment]
> Have you tried OneHotEncoder in scikit-learn?
> https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html#sklearn.preprocessing.OneHotEncoder
> If that works, adding an option at the position below seems to be an simple solution.
> https://github.com/bgruening/galaxytools/blob/master/tools/sklearn/main_macros.xml#L788

@qiagu Yes, that works too. to_categorical() allows you to specify the number of categories (which OneHotEncoder does not). So, it seems to_categorize is a bit more flexible.

**Comment by bgruening** [comment]
Does anyone know what that means?

```
Traceback (most recent call last):
  File "/tmp/tmpn5y431yl/job_working_directory/000/24/configs/tmpzpix3cx1", line 11, in <module>
    from galaxy_ml.utils import read_columns, SafeEval
  File "/usr/local/lib/python3.6/site-packages/galaxy_ml/utils.py", line 15, in <module>
    import xgboost
  File "/usr/local/lib/python3.6/site-packages/xgboost/__init__.py", line 11, in <module>
    from .core import DMatrix, Booster
  File "/usr/local/lib/python3.6/site-packages/xgboost/core.py", line 136, in <module>
    _LIB = _load_lib()
  File "/usr/local/lib/python3.6/site-packages/xgboost/core.py", line 128, in _load_lib
    lib.XGBGetLastError.restype = ctypes.c_char_p
UnboundLocalError: local variable 'lib' referenced before assignment
```

**Comment by anuprulez** [comment]
Looks like `xgboost` is not loading properly.

**Comment by kxk302** [comment]
Fixed a bug after local testing and removed print statements.

File: tools/sklearn/to_categorical.xml [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+tf.keras.utils.to_categorical(
+    y, num_classes=None, dtype='float32'
+)
+
+E.g. for use with categorical_crossentropy.
+
+Arguments
+
+y: class vector to be converted into a matrix (integers from 0 to num_classes).
```
Comment (bgruening) [inline tools/sklearn/to_categorical.xml:79]: maybe that is too technical?

Comment (kxk302) [reply tools/sklearn/to_categorical.xml]: Simplified the description a bit.

File: tools/sklearn/to_categorical.xml [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+tf.keras.utils.to_categorical(
+    y, num_classes=None, dtype='float32'
+)
+
+E.g. for use with categorical_crossentropy.
+
+Arguments
+
+y: a vector of numbers to be converted into a vector of one-hot encoded values.
```
Comment (anuprulez) [inline tools/sklearn/to_categorical.xml:79]: the transformation should be a matrix, correct?

Comment (kxk302) [reply tools/sklearn/to_categorical.xml:79]: Yes, will revise.

**Comment by bgruening** [comment]
> Files /tmp/tmpy5vu3vs1model_fit02=1136b but /tmp/tmpqechsbtxmodel_fit02=1167b - compare by size (delta=30) failed

Fo a few the diff of 30 is not enough.

--------

The non-writeable error looks very wired, I will try to look at it.

-------------

```
Traceback (most recent call last):
  File "/home/runner/.planemo/planemo_tmp_4u8hwhsx/train_test_eval.py", line 434, in <module>
    targets=args.targets, fasta_path=args.fasta_path)
  File "/home/runner/.planemo/planemo_tmp_4u8hwhsx/train_test_eval.py", line 405, in main
    del main_est.validation_data
AttributeError: validation_data
```

This one is realy one introduced by some updates I think.

So the good news is, its only those 3 errors :)

File: tools/sklearn/main_macros.xml [hunk 4]
```diff
@@ -3,8 +3,7 @@
 
   <xml name="python_requirements">
       <requirements>
-          <requirement type="package" version="3.6">python</requirement>
-          <requirement type="package" version="0.8.1">Galaxy-ML</requirement>
+          <requirement type="package" version="0.8.3">Galaxy-ML</requirement>
```
Comment (kxk302) [inline tools/sklearn/main_macros.xml:6]: @bgruening I noticed that the requirement for python3.6 is removed. I put it back (locally) and ran the tests again, and I got the "UnboundLocalError: local variable 'lib' referenced before assignment" error again. I am assuming (please confirm) that the python3.6 requirement is needed (and I deleted it by mistake). Anyhow, the test errors change depending on having or not having this requirement.

Comment (bgruening) [reply tools/sklearn/main_macros.xml:6]: It should not be needed as Galaxy-ML depends on python, so Galaxy-ML will pull in python.

File: tools/sklearn/keras_train_and_eval.xml [hunk 5]
```diff
@@ -19,30 +19,27 @@
...
-            --outfile_y_true '$outfile_y_true'
-            --outfile_y_preds '$outfile_y_preds'
+                --outfile_y_true '$outfile_y_true'
+                --outfile_y_preds '$outfile_y_preds'
             #end if
             #if $experiment_schemes.test_split.split_algos.shuffle == 'group'
-            --groups '$experiment_schemes.test_split.split_algos.groups_selector.infile_g'
+                --groups '$experiment_schemes.test_split.split_algos.groups_selector.infile_g'
             #end if
-            >'$outfile_result' && cp '$outfile_result' "`pwd`/../tool_stdout"
```
Comment (bgruening) [inline tools/sklearn/keras_train_and_eval.xml:43]: @kxk302 this is the error. This is really not good and was not caught without the container tests. So its good we have these more stringent tests.

File: tools/sklearn/main_macros.xml (hunk 4 above)
Comment (kxk302) [reply tools/sklearn/main_macros.xml:6]: Ok, thats good.

**Comment by bgruening** [comment]
@kxk302 down to 8 errors out of 231 tests :) Not bad.

Do you want to care about the rest?

**Comment by kxk302** [comment]
> @kxk302 down to 8 errors out of 231 tests :) Not bad.
> 
> Do you want to care about the rest?

Yes, I will! Thx  @bgruening.

**Comment by bgruening** [comment]
Nice, all green besides the linting.

@kxk302 let me know if this should be merged.

**Comment by kxk302** [comment]
> Nice, all green besides the linting.
> 
> @kxk302 let me know if this should be merged.

@bgruening yes, please merge.

File: tools/sklearn/train_test_eval.py [hunk 6]
```diff
@@ -395,23 +420,25 @@ def main(inputs, infile_estimator, infile1, infile2,
         if isinstance(estimator, pipeline.Pipeline):
```
Comment (bgruening) [inline tools/sklearn/train_test_eval.py:420]: @kxk302 pipeline does not seem to be defined? At least the linter complains and this seems to be serious.

Comment (kxk302) [reply tools/sklearn/train_test_eval.py:424]: @bgruening Hmm, I had started on fixing some of the lint issues and it seems I did not push some of the files. Pulled from upstream and pushed them just now with revised files. When tests are green, it should be good to go. Thanks.

**Comment by kxk302** [comment]
I see some test failures. I will fix these tomorow.

File: tools/sklearn/fitted_model_eval.py [hunk 7]
```diff
@@ -103,49 +106,55 @@ def main(inputs, infile_estimator, outfile_eval,
...
+    # Check if scoring is specified
+    scoring = params["scoring"]
+    if scoring is not None:
+        # get_scoring() expects secondary_scoring to be a comma separated string (not a list)
+        # Check if secondary_scoring is specified
+        secondary_scoring = scoring.get("secondary_scoring", None)
+        if secondary_scoring is not None:
+            # If secondary_scoring is specified, convert the list into comman separated string
+            scoring["secondary_scoring"] = ",".join(scoring["secondary_scoring"])
+
```
Comment (qiagu) [inline tools/sklearn/fitted_model_eval.py:140]: Galaxy used to generate comma separated string for multiple select options. Has this changed? Did the change start from galaxy version 20.05? If not, do we need a branch checker here for backward compatibility? `get_scoring()` is called at multiple places, we'd better change it in utils for future releases.

File: tools/sklearn/fitted_model_eval.py [hunk 8]
```diff
@@ -103,49 +106,55 @@ def main(inputs, infile_estimator, outfile_eval,
...
     scorer = get_scoring(scoring)
     scorer, _ = _check_multimetric_scoring(estimator, scoring=scorer)
 
-    if hasattr(estimator, 'evaluate'):
-        scores = estimator.evaluate(X_test, y_test=y_test,
-                                    scorer=scorer,
-                                    is_multimetric=True)
+    if hasattr(estimator, "evaluate"):
+        scores = estimator.evaluate(X_test, y_test=y_test, scorer=scorer, is_multimetric=True)
```
Comment (qiagu) [inline tools/sklearn/fitted_model_eval.py:145]: @kxk302  What's the maximum line length? We can make it longer, but I do think we still need a number.

File: tools/sklearn/fitted_model_eval.xml [hunk 9]
```diff
@@ -1,11 +1,11 @@
-<tool id="sklearn_fitted_model_eval" name="Evaluate a Fitted Model" version="@VERSION@">
+<tool id="sklearn_fitted_model_eval" name="Evaluate a Fitted Model" version="@VERSION@" profile="20.05">
```
Comment (qiagu) [inline tools/sklearn/fitted_model_eval.xml:1]: Any particular reason for targeting profile '20.05'?

Comment (bgruening) [reply tools/sklearn/fitted_model_eval.xml:1]: we can also set 20.01, but we should set one. Galaxy will treat $HOME etc differently with a set profile.

File: tools/sklearn/fitted_model_eval.py (hunk 7 above)
Comment (bgruening) [reply tools/sklearn/fitted_model_eval.py:140]: I don't think this has changed. Hopefully not.

File: tools/sklearn/train_test_eval.xml [hunk 10]
```diff
@@ -1,11 +1,11 @@
-<tool id="sklearn_train_test_eval" name="Train, Test and Evaluation" version="@VERSION@">
+<tool id="sklearn_train_test_eval" name="Train, Test and Evaluation" version="@VERSION@" profile="20.05">
```
Comment (qiagu) [inline tools/sklearn/train_test_eval.xml:1]: Do you think this tool is still needed? I thought it was deprecated a while ago.

**Comment by qiagu** [comment]
@kxk302  Thanks a lot for making this PR. It must take a lot of effort to accomplish this. I appreciate most of the changes, while the enforcement of double quotes in python code, IMHO, may not be necessary, unless python published a new guidance for that. I basically followed a rule in which single quotes were for key words, double quotes for human readable sentences. However, the quotes are interchangeable from what I read. Since you have made them, I can use double quotes as well. Again, this is a great PR. Thanks!

**Comment by bgruening** [comment]
@qiagu I think that was me. Sorry there has been many flake8 warnings and I just used https://github.com/psf/black to convert the files.

File: tools/sklearn/fitted_model_eval.py (hunk 7 above)
Comment (kxk302) [reply tools/sklearn/fitted_model_eval.py:140]: I've added a (temporary) fix to galaxytools. Will add a fix to galaxy-ml afterwards (and take out the fix in galaxytools).

File: tools/sklearn/train_test_eval.xml (hunk 10 above)
Comment (bgruening) [reply tools/sklearn/train_test_eval.xml:1]: I don't have a feeling for this. @anuprulez ?

**Comment by qiagu** [comment]
Okay, the black sounds great to me.

**Comment by kxk302** [comment]
@bgruening The tests are all passing (minus the one commented out). Could you please merge this when you have time? Thanks.

**Comment by bgruening** [comment]
Than lets merge it :)

**Comment by kxk302** [comment]
This is the ticket to address the commented out test: https://github.com/bgruening/galaxytools/issues/1102

**Comment by bgruening** [comment]
Thanks @kxk302!

**Comment by kxk302** [comment]
@bgruening should I create a PR to have this installed in galaxy EU? This link says a PR is needed for new tools, but I'm not sure about existing tools.
https://galaxyproject.eu/posts/2020/08/22/three-steps-to-galaxify-your-tool/

**Comment by bgruening** [comment]
Yes, such a PR would be needed. Thanks!

**Comment by kxk302** [comment]
@bgruening here is the PR: https://github.com/usegalaxy-eu/usegalaxy-eu-tools/pull/377

**Pull request merged**